core/src/main/java/hudson/model/Job.java

```diff
@@ -684,9 +685,37 @@ public void movedTo(DirectlyModifiableTopLevelItemGroup destination, AbstractIte
         }
     }
 
-    @Override public void delete() throws IOException, InterruptedException {
+    @Override
+    public void delete() throws IOException, InterruptedException {
         super.delete();
         Util.deleteRecursive(getBuildDir());
+        if (this instanceof TopLevelItem) {
+            cleanupWorkspaces((TopLevelItem) this);
+        }
+    }
+
+    /**
+     * Clean out workspaces for attached agents (best efforts, as they may not always be online)
+     * There is also hudson.model.WorkspaceCleanupThread to catch ones that were missed.
+     */
+    private void cleanupWorkspaces(TopLevelItem item) {
+        Jenkins j = Jenkins.getInstance();
+        List<Node> nodes = new ArrayList<>();
+        nodes.add(j);
+        nodes.addAll(j.getNodes());
+        Logger log = Logger.getLogger(Job.class.getName());
+        for (Node node : nodes) {
+            FilePath ws = node.getWorkspaceFor(item);
```

Comment:
@jtnord @olivergondza @batmat agree - this is also my concern. 

One option is to run the node deletions in the background, but not the delete of the master directory (as it preserves current behaviour). For git-flowy like cases where there are lots of things to delete, generally the directories are smaller (at least on the master) as they have been much shorter lived.